File: tools/simurgh-daemon-macos/Sources/SimurghDaemon/ProofSigner.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
import Foundation

final class ProofSigner {
    let identity: KeychainIdentity
    let scanner: AffinityScanner

    init(identity: KeychainIdentity, scanner: AffinityScanner = AffinityScanner()) {
        self.identity = identity
        self.scanner = scanner
    }

    func pair(sessionId: String, examId: String, challenge: String) throws -> [String: Any] {
        let payload: [String: Any] = [
            "type": "simurgh.daemon.pair",
            "session_id": sessionId,
            "exam_id": examId,
            "challenge": challenge,
            "timestamp": isoNow(),
            "node_id_hash": identity.nodeIdHash,
            "daemon_version": "0.4.7",
            "platform": "macos",
        ]
        return [
            "ok": true,
            "node_id_hash": identity.nodeIdHash,
            "public_key": identity.publicKey,
            "signature": try identity.sign(payload),
            "signed_payload": payload,
        ]
    }

    func proof(sessionId: String, examId: String, sequence: Int, challenge: String) throws -> [String: Any] {
        let scan = scanner.scan()
        var payload: [String: Any] = [
            "type": "simurgh.daemon.proof",
            "session_id": sessionId,
            "exam_id": examId,
            "sequence": sequence,
            "timestamp": isoNow(),
            "node_id_hash": identity.nodeIdHash,
            "daemon_version": "0.4.7",
            "platform": "macos",
            "helper_state": scan.scannerState == "risk_detected" ? "risk_detected" : "healthy",
            "challenge": challenge,
        ]
        for (key, value) in scan.asDictionary() {
            payload[key] = value
        }
        payload["signature"] = try identity.sign(payload)
        return ["ok": true, "daemon_proof": payload]
    }
}

File: tools/simurgh-daemon-windows/src/SimurghDaemon.Windows/ProofSigner.cs
using System.Security.Cryptography;
using System.Text;

namespace SimurghDaemon.Windows;

public sealed class ProofSigner(WindowsIdentityStore identity)
{
    public DaemonProof CreateProof(
        string sessionId,
        string examId,
        int sequence,
        string challenge,
        AffinityScanResult scan,
        DateTimeOffset timestamp)
    {
        var unsigned = new DaemonProof(
            Type: "simurgh.daemon.proof",
            SessionId: sessionId,
            ExamId: examId,
            Sequence: sequence,
            Timestamp: timestamp.UtcDateTime.ToString("O"),
            NodeIdHash: identity.NodeIdHash,
            DaemonVersion: "0.4.11",
            Platform: "windows",
            ScannerState: scan.ScannerState,
            ScannerVersion: scan.ScannerVersion,
            ScanTimestamp: timestamp.UtcDateTime.ToString("O"),
            ScanDurationMs: 0,
            VisibleWindowCount: scan.VisibleWindowCount,
            SuspiciousWindowCount: scan.SuspiciousWindowCount,
            CaptureExcludedWindowCount: scan.CaptureExcludedWindowCount,
            CaptureRestrictedWindowCount: scan.CaptureRestrictedWindowCount,
            MonitorOnlyWindowCount: scan.MonitorOnlyWindowCount,
            ScanErrorCount: scan.ScanErrorCount,
            PrivacyMode: scan.PrivacyMode,
            WindowFingerprintHashes: [],
            HelperState: "healthy",
            Challenge: challenge,
            Signature: "");
        var canonical = CanonicalJson.SerializeWithoutSignature(unsigned);
        var signature = identity.Key.SignData(Encoding.UTF8.GetBytes(canonical), HashAlgorithmName.SHA256, DSASignatureFormat.Rfc3279DerSequence);
        return unsigned with { Signature = Base64Url.Encode(signature) };
    }

    public bool Verify(DaemonProof proof)
    {
        try
        {
            var signature = Convert.FromBase64String(proof.Signature.Replace('-', '+').Replace('_', '/').PadRight(proof.Signature.Length + (4 - proof.Signature.Length % 4) % 4, '='));
            var canonical = CanonicalJson.SerializeWithoutSignature(proof);
            return identity.Key.VerifyData(Encoding.UTF8.GetBytes(canonical), signature, HashAlgorithmName.SHA256, DSASignatureFormat.Rfc3279DerSequence);
        }
        catch
        {
            return false;
        }
    }

    public DaemonPairResponse CreatePair(
        string sessionId,
        string examId,
        string challenge,
        DateTimeOffset timestamp)
    {
        var signedPayload = new DaemonPairSignedPayload(
            Type: "simurgh.daemon.pair",
            SessionId: sessionId,
            ExamId: examId,
            Challenge: challenge,
            Timestamp: timestamp.UtcDateTime.ToString("O"),
            NodeIdHash: identity.NodeIdHash,
            DaemonVersion: "0.4.11",
            Platform: "windows");
        var canonical = CanonicalJson.SerializeWithoutSignature(signedPayload);
        var signature = identity.Key.SignData(
            Encoding.UTF8.GetBytes(canonical),
            HashAlgorithmName.SHA256,
            DSASignatureFormat.Rfc3279DerSequence);
        return new DaemonPairResponse(true, identity.NodeIdHash, identity.PublicKey, signedPayload, Base64Url.Encode(signature));
    }
}
